Support timed transactions

ffakenz opened the pull request:
Depends on: #862 

This is a first step to fix #196, focused on the **transaction validation** part of this issue.

🌴 The internal Ledger now uses a ChainSlot during tx validation.
This is used to configure the SlotNo ledger env setting before validation.

🌴 Add the notion of current slot in the head while in open state.
What evolves time on the L1 is what evolves time on L2, eg. observing blocks.
This is obtained from new added slot number field to the Tick chain event.
This slot is used when validating incoming tx (from ReqTx and ReqSn network events) against the internal ledger.

🌴 In this first ("dirt road") solution, we ignore finer grained slots/time resolution.

🌴 Just like on the exercised e2e spec, Clients are expected to just use "wall clock time" to define their validity range, possibly using some client-side API to simplify the conversion to slots (can use cardano-cli, or most client SDKs). They are supposed to know which (L1) network their Head is opened on

<img width="1428" alt="image" src="https://github.com/input-output-hk/hydra/assets/14299604/8adc5f20-2fa6-4408-98f5-886ca6961bdf">

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Add e2e spec for timed txs ~ first draft

This excercises the lack of support to validate txs
which are valid at some slot no in the future.

Also added helper function to create txs with validity range.

ffakenz committed:
Add notion of current slot while in open state.

This is obtained from new added slot number to Tick chain event.

This slot will be used later to validate tx against.

ffakenz committed:
Validate timed txs using current slot in open head

The internal Ledger now uses a ChainSlot during tx validation.
This is used to configure the SlotNo ledger env setting before
validation.

Add test to excercise prune of invalid transactions during ReqSn validation.
It generates transactions which are going to expire some slot in the future.

ChainSlot was moved from module due to cyclic dependency errors.

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/ChainPoint.hs:
```diff
@@ -11,6 +11,12 @@ getChainPoint header =
 
+-- | Get the chain SlotNo corresponding to a given 'BlockHeader'.
+getChainSlotNo :: BlockHeader -> SlotNo
+getChainSlotNo header = slotNo
```
Should: Move this to a new `SlotNo` module or at least the `BlockHeader` module.

There is a convention where to put functions in `hydra-cardano-api`. See https://github.com/input-output-hk/hydra/tree/master/hydra-cardano-api#organization--naming-conventions

ch1bo replied:
That being said.. this function is not needed. See my other comment in `Handler.hs`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -245,20 +246,23 @@ chainSyncHandler tracer callback getTimeHandle ctx localChainState =
-            throwIO TimeConversionException{slotNo, reason}
-          Right utcTime ->
-            callback (Tick utcTime)
+    let slotNo = getChainSlotNo header
```
No need to extract the slot from the block header (we can drop `getChainSlotNo` if you don't want to put it into the right module / name it consistent). We have it already with the `chainSlot` below.

You also could have just taken the code from before and

```
callback Tick{chainTime = utcTime, chainSlot = ChainSlot . fromIntegral $ unSlotNo slotNo}
```

or a similar conversion.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -87,9 +88,9 @@ mkTimeHandle currentSlotNo systemStart eraHistory = do
+  slotToUTCTime :: SlotNo -> Either Text UTCTime
   slotToUTCTime slot =
-    case interpretQuery interpreter (slotToWallclock slot) of
+    case getProgress slot eraHistory of
```
Interesting, didn't know this existed.

However, as long as there is no equivalent for the "other direction" used below, I would argue it's better to depend on `ouroboros-consensus` here and use the function from there.

Should: revert or find a way to get rid of `import Ouroboros.Consensus.HardFork.History.Qry`

ch1bo commented on hydra-node/src/Hydra/Ledger.hs:
```diff
@@ -64,9 +66,22 @@ data Ledger tx = Ledger
+nextChainSlot (ChainSlot n) = ChainSlot (n + 1)
+
+instance Arbitrary ChainSlot where
+  arbitrary = genericArbitrary
```
Should: Move the `ChainSlot` its instances and `nextChainSlot` not between `Ledger` and `canApply` definitions.

They belong together and we moved it here in between. Let's keep the things cohesive (if you want.. even in a dedicated module).

ffakenz replied:
Moved the `ChainSlot` type and related functions above `Ledger`.

We can move it to its own module if you want, just let me know.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Builder.hs:
```diff
@@ -55,23 +55,24 @@ instance Exception InvalidTransactionException
+    , txUpdateProposal = TxUpdateProposalNone
+    , txMintValue = TxMintValueNone
+    , txScriptValidity = TxScriptValidityNone
+    }
```
Thanks, but would have not been required in this PR :)

ffakenz replied:
I know, but is smth i've been wanting to push for a long time :P

ch1bo commented on hydra-node/src/Hydra/Ledger/Simple.hs:
```diff
@@ -111,7 +111,7 @@ instance FromCBOR SimpleTxIn where
 simpleLedger =
   Ledger
-    { applyTransactions =
+    { applyTransactions = \slot ->
```
This is currently unused and makes CI fail. It's fine to mark this as `_slot` and document that `SimpleTx` transactions don't have a notion of time.

I realize that the module documentation above is severly outdated. Maybe we can use the chance to update it to say what `SimpleTx` is and what it is not (i.e. no time validity)

ffakenz replied:
I updated the documentation but I did not find it outdated.

Can you plz give a look to my changes and describe which updates would you introduce to the module description?

ch1bo replied:
The module haddock refers to a `Mock` transaction which does not exist anymore. I would propose to write something like:

```
-- | A mock implementation of a ledger using very simple UTxO transactions.
--
-- These transactions have a very simplified representation of unspent transaction outputs being just integers, but already have inputs and outputs. Transactions are validated against the current state of the ledger, so that
--   one transaction could at some point be invalid, then becomes valid because some inputs it
-- consumes is now available.
--
-- NOTE: There is no notion of time in this ledger, so transactions validation will never depend on the L1 slot.
```

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -606,10 +605,11 @@ simulatedChainAndNetwork initialChainState = do
     threadDelay blockTime
     now <- getCurrentTime
-    readTVarIO nodes >>= \ns -> mapM_ (`handleChainEvent` Tick now) ns
+    readTVarIO nodes >>= \ns -> mapM_ (`handleChainEvent` Tick now (chainStateSlot cs)) ns
```
This is smelly. Don't read a TVar (`chainStateVar` here) and then wait.. it might have changed until you use this value (unless you deliberately want to take the value before waiting, but I doubt this is what we want here).

In general, it's always a good idea to group multiple STM actions into a single `atomically` block to ensure things are consistent, especially as we are writing to `chainStateVar` from other threads!

Must: read chain state just before calling back with that slot

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/HandlersSpec.hs:
```diff
@@ -102,7 +106,7 @@ spec = do
             either (failure . ("Time conversion failed: " <>) . toString) pure $
               slotToUTCTime timeHandle slot
-        void . stop $ events === [Tick expectedUTCTime]
+        void . stop $ events === [Tick expectedUTCTime (ChainSlot . fromIntegral . unSlotNo $ slot)]
```
Could: move $ to avoid a .

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -279,7 +286,7 @@ spec = do
         s1 <- assertNewState outcome1
         let oneSecondsPastDeadline = addUTCTime 1 contestationDeadline
-            stepTimePastDeadline = OnChainEvent $ Tick oneSecondsPastDeadline
+            stepTimePastDeadline = OnChainEvent $ Tick oneSecondsPastDeadline (ChainSlot 1)
```
Could: use a weirder number than 1 or ``someSlot = arbitrary `generateWith` 42`` to indicate it should not matter which we choose

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -313,6 +320,54 @@ spec = do
         s1 `hasEffect` contestTxEffect
         assertOnlyEffects s1
 
+    describe "Coordinated Head Protocol using real Tx" $
```
Could: move this one level down and use the fixtures (`threeParties` and `bobEnv`) of the outer describe instead of defining `parties` and `env` here again (both do not really matter for this test)

ffakenz replied:
Moved the fixtures (`threeParties` and `bobEnv`) to an outer scope and reuse them here.

ch1bo commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -10,17 +10,7 @@ import Hydra.API.ClientInput (ClientInput (..))
-  PostChainTx (InitTx),
-  PostTxError (NoSeedInput),
- )
+import Hydra.Chain (Chain (..), ChainEvent (..), HeadId (HeadId), HeadParameters (HeadParameters), IsChainState, OnChainTx (..), PostChainTx (InitTx), PostTxError (NoSeedInput))
```
Should: not inline lines which are too long. Just keep it the way it was?

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -207,7 +194,15 @@ data ChainEvent tx
+    -- the alternative would be to have the mans to do the conversion.
+    -- For Cardano, this would be a systemStart and eraHistory..
+    -- which is annoying and if it's kept in the chain layer,
+    -- it would mean another round trip / state to keep there.
```
```suggestion
  | -- | Indicate time has advanced on the chain.
    --
    -- NOTE: While the type does not guarantee that the UTCTime and ChainSlot
    -- are consistent the alternative would be provide the means to do the
    -- conversion. For Cardano, this would be a systemStart and eraHistory..
    -- which is annoying and if it's kept in the chain layer, it would mean
    -- another round trip / state to keep there.
```

ch1bo replied:
```suggestion
  | -- | Indicate time has advanced on the chain.
    --
    -- NOTE: While the type does not guarantee that the UTCTime and ChainSlot
    -- are consistent the alternative would be provide the means to do the
    -- conversion. For Cardano, this would be a systemStart and eraHistory..
    -- which is annoying and if it's kept in the chain layer, it would mean
    -- another round trip / state to keep there.
```

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1511,6 +1511,12 @@ components:
+      type: integer
+      minimum: 0
+      description: |
+        A number representing the time periods (slots) progressed on the chain.
```
```suggestion
        A number representing discrete time (slots) on the chain.
```

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -121,6 +126,11 @@ spec = around showLogsOnFailure $ do
             publishHydraScriptsAs node Faucet
               >>= canCloseWithLongContestationPeriod tracer tmpDir node
+      it "can submmit a timed tx" $ \tracer -> do
+        withClusterTempDir "timmed-tx" $ \tmpDir -> do
```
```suggestion
        withClusterTempDir "timed-tx" $ \tmpDir -> do
```

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,76 @@ waitForLog delay nodeOutput failureMessage predicate = do
+
+    -- Acquire a current point in time
+    now' <- getCurrentTime
+    let slotLengthMillis = 100
```
Must: make this not hard-coded

This is the last mile as far as I see. We should query / infer this from the network we are connected to. If we are not able to do this here (are we running on the devnet with 0.1s or on the testnet with 1s slot length?), users will neither be able to do this.

Can we query this from the node? If not, we can look it up from the config when starting the cardano-node/network and store it the `RunningNode` like the network id.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,76 @@ waitForLog delay nodeOutput failureMessage predicate = do
+
+    -- Create an arbitrary transaction using some input.
+    let paymentFromAliceToFaucet :: Num a => a
+        paymentFromAliceToFaucet = 1_000_000
```
Should: avoid this variable.

It's only used in one location, no need for a symbol just for this. Also, this must correspond to funds we originally paid to alice. If we want to make this more robust, we should rather put the `100_000_000` above into a variable and use that when determining the value sent on line 548

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,76 @@ waitForLog delay nodeOutput failureMessage predicate = do
+    let paymentFromAliceToFaucet :: Num a => a
+        paymentFromAliceToFaucet = 1_000_000
+    let firstCommittedUTxO = Prelude.head $ UTxO.pairs committedUTxOByAlice
+    (faucetVk, _) <- keysFor Faucet
```
Could: simplify the transaction further.

Have alice pay funds to herself would be even more simple and not require this statement for example.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,76 @@ waitForLog delay nodeOutput failureMessage predicate = do
+
+      (ChainSlot nbr) = ChainSlot 0
+      slotsSince = SlotNo . truncate $ (diffUTCTime now' now / slotLengthSec)
+      currentSlot = SlotNo (fromIntegral nbr) + slotsSince
```
This is weird.. are you checking the difference between starting the test and now to find out what the current slot would be?

I think this is also very fragile and it will never work on the testnet as this was there before we started this test.

Instead, we need to either query the current slot from the cardano-node or use the wall clock time together with a known start time of the network (we can query that as well).

ch1bo reviewed (changes requested):
Looks quite good already! The core changes seem to have been done and this mostly requires some cleanup and the e2e test to exercise our stance here:

Clients must know the current slot and slot length by the network the Head is currently running on.

So let's demonstrate how that is done!

ffakenz committed:
Add note comment on Tick chain event

ffakenz committed:
Derive chainSlot from slotNo instead of point

This makes the time conversion more obvious and less obfuscated.

ffakenz committed:
Fix chain state var handling on behavior spec.

This was smelly.
We should don't read a TVar (chainStateVar here) and then wait,
as it might have changed until you use this.

This fix makes the event consistent, as chain slot is consistent with the current time queried before.

ffakenz committed:
Add fixtures to be used across different coordinated head protocol tests

ffakenz committed:
Improve api description for ChainSlot

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Fix typo on tmp dir name

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Fix unused-matches warn on Simple module

Include notes about SimpleTx transactions don't supporting time validity ranges.

Also took the opportunity to apply some minor hlint improvements

ffakenz committed:
Make use of cardano client to query shelly parameters on e2e spec

This is to make the test more robust by removing hardcoded values.

ffakenz committed:
Reorder types in Ledger to improve module  cohesiveness

ch1bo committed:
Do not convert to ShelleyGenesis

We can just use the cardano-api function to get the slot length

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,70 @@ waitForLog delay nodeOutput failureMessage predicate = do
+
+    -- Create a transaction which is only valid in 5 seconds
+    let
+      secondsToAwait = 15
```
Should: fix comment above or this literal, 5 /= 15

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -487,6 +497,70 @@ waitForLog delay nodeOutput failureMessage predicate = do
+      guard $ v ^? key "tag" == Just "TxInvalid"
+
+    -- Wait for the future chain slot and time
+    threadDelay 15
```
Should: use the variable from above to wait for the right amount of time

Also.. 15 seconds is a bit long for a happy path execution

ch1bo commented on hydra-node/src/Hydra/Ledger/Simple.hs:
```diff
@@ -20,14 +20,20 @@ import Data.Aeson (
  )
```
Should: Fix the module haddock above with the text I provided last week.

ch1bo reviewed (approved):
Only minor things left, let's get this done today :)

ffakenz committed:
Update Simple module description

ffakenz committed:
Decrease seconds to wait in e2e spec to 5

Also reuse same value during threadDelay.

Note: the time package was included to hydra-cluster to perform time operations.

ffakenz committed:
Update CHANGELOG

GeorgeFlerovsky commented:
Thanks :1st_place_medal:

uhbif19 commented:
@ch1bo @ffakenz

Seems like Tick is only introduced in logs, but not in server outputs.

But we will probably need it to create TimeHandle.

ch1bo commented:
@uhbif19 We deliberately did not add any "publishing of time" part in this first PR as we need more information on what would be needed there and didn't want to delay this work so far. See also [here](https://github.com/input-output-hk/hydra/issues/196#issuecomment-1540045846). If you are using the `TimeHandle` api from our code base, you would want to query it from the cardano-node.

Now.. what would you expect to have in a "time API" and how often should it be sent by the hydra-node?

uhbif19 commented:
> you are using the TimeHandle api from our code base, you would want to query it from the cardano-node.

Yes, that is what we gonna do for now. But seems awkward to use L1 connection in L2 client/monad.

> Now.. what would you expect to have in a "time API" and how often should it be sent by the hydra-node?

I am not sure that there is any choice. Any L2 slot/tip change should be available. The only question I see is if it should be pushed by server on every tip advance or queried by client.
